src/couch/priv/couch_js/60/util.cpp

```diff
@@ -30,36 +32,32 @@ js_to_string(JSContext* cx, JS::HandleValue val)
     JS::UniqueChars chars(JS_EncodeStringToUTF8(cx, sval));
     if(!chars) {
         JS_ClearPendingException(cx);
-        fprintf(stderr, "Error converting value to string.\n");
-        exit(3);
+        return std::string();
     }
 
     return chars.get();
 }
 
-std::string
-js_to_string(JSContext* cx, JSString *str)
+JSString*
+string_to_js(JSContext* cx, const std::string& raw)
 {
-    JS::UniqueChars chars(JS_EncodeString(cx, str));
-    if(!chars) {
-        JS_ClearPendingException(cx);
-        fprintf(stderr, "Error converting  to string.\n");
-        exit(3);
+    JS::UTF8Chars utf8(raw.c_str(), raw.size());
+    JS::UniqueTwoByteChars utf16;
+    size_t len;
+
+    utf16.reset(JS::UTF8CharsToNewTwoByteCharsZ(cx, utf8, &len).get());
+    if(!utf16) {
+        return nullptr;
     }
 
-    return chars.get();
-}
+    JSString* ret = JS_NewUCString(cx, utf16.get(), len);
 
-JSString*
-string_to_js(JSContext* cx, const std::string& s)
-{
-    JSString* ret = JS_NewStringCopyN(cx, s.c_str(), s.size());
-    if(ret != nullptr) {
-        return ret;
+    if(ret) {
+        // JS_NewUCString took ownership on succses
+        mozilla::Unused << utf16.release();
```

Comment:
Not so sure about this, and @jiangphcn saw a double free in CI output.

Reply:
double free in CI output with this? Or from the other thing we talked about?

Reply:
I pulled that from SpiderNode to be clear. I'm not 100% on the magic of that other than the ownership makes sense to me.

Reply:
Looking at SearchFox suggests that `mozilla::Unused` is only to suppress GCC warnings about unused return values:

https://searchfox.org/mozilla-central/source/mfbt/Unused.cpp
https://searchfox.org/mozilla-central/source/mfbt/Unused.h#17-24

I see evidence of it in the current tree for `js.cpp`:

https://searchfox.org/mozilla-central/source/js/src/shell/js.cpp#6041

so this is probably OK.

However, this might be a more elegant pattern:

https://searchfox.org/mozilla-central/source/js/src/vm/CompilationAndEvaluation.cpp#146-152

Note the `.get()` directly in the ctor for `JS:UniqueTwoBytesChars`. You could do this directly on line 46 and avoid the `.reset()` call. (I only used `.reset()` in one place to try and preserve the wacky `goto` hijinx in my PR, but I think in the end I removed the `goto` and stopped trying to work around the `goto` crossing the initialization error.

Double free might be related, unsure. Has it happened again?

Reply:
Ah, I should comment on that. Yeah, the JS/Mozilla headers have a lot of declarations about forcing errors and warnings so that's specifically to silence the warning (which makes sense cause without knowing where that pointer went it looks like a leak).

I stole the pattern from SpiderNode who've been using it unchanged for multiple years. For the second version where its like three levels deep in the constructor I did it this way to avoid that style eyesore. I had to read it like five times to even get an idea of what they were doing.

I'll add a comment about the Unused bit being used to silence the compiler.